### PR TITLE
feat: Autodetect npm.packageManager #102050

### LINKED ISSUE #102050: Automatically detect package manager (npm.packageManager)
# Problem
When switching between many JavaScript projects, it's often easy to forget which package manager should be used in each project. JavaScript package managers aren't quite compatible either. Each package manager resolves dependencies differently, so accidentally installing with npm could cause a yarn project to break.

While VSCode has the `npm.packageManager` option to switch between npm, yarn, and pnpm, it's not very convenient to have to remember to set it on every project. Even if a user has a global preference, they will often have to clone several projects they're contributing to using other package managers anyway. Personally, I have so many projects installed that it's too inconvenient for me to set this option, so I end up figuring out which package manager to use manually and running it in the terminal.

# Proposed option changes
I would like `npm.packageManager` to default to a new `auto` option, which would select a package manager based on the presence of workspace and lock files using [`preferred-pm`](https://www.npmjs.com/package/preferred-pm). If a lock file doesn't exist, the user is probably using npm anyway, and we can just default to `npm` like we already are. It would also be nice to set up some sort of warning if multiple lock files are found, as most package managers do this to inform users that their package installations can break because of differences in version resolution.

# Ideas for extension API changes
I'm not actually familiar with the extension APIs, but I think it would be nice to expose this as some sort of generic API for extensions so they don't have to reimplement this functionality or reinstall dependencies. This would also prevent extensions from breaking if VSCode changed its package manager detection heuristics in the future. This also might be useful for other internal features like automatically installing projects when cloned with the right package manager, though with the extension API users would have an alternative if something wasn't officially supported.

### PR DESCRIPTION
**Pull request by shivangg**
This PR fixes #102050 

- [x]  `npm.packageManager` should have `auto` option **and** default to it.
- [x]  When `auto`, check lock file to find preferred package manager.
- [x] Warn if `auto` set and multiple lock files found.

### CODE REVIEW

**Comment by shivangg** [comment]
Hey @alexr00 One of the check is waiting to be reported. Can you help me understand the reason for that. 

Also, is this blocking the PR review?

**Comment by shivangg** [comment]
@alexr00 Could you please advise an update on this PR?

**Review by nickserv — APPROVED** [review]

**Comment by nickmccurdy** [comment]
Looks like there's a merge conflict in `tasks.ts`

**Comment by shivangg** [comment]
@nickmccurdy @alexr00  Resolved the merge conflict.

**Review by nickserv — APPROVED** [review]

**Comment by shivangg** [comment]
@alexr00 Is something pending on this PR?  I'd love to resolve it.

File: extensions/npm/src/tasks.ts [hunk 1]
```diff
@@ -107,8 +108,20 @@ export function isWorkspaceFolder(value: any): value is WorkspaceFolder {
...
-	return workspace.getConfiguration('npm', folder.uri).get<string>('packageManager', 'npm');
+export async function getPackageManager(folder: WorkspaceFolder): Promise<string> {
+	let packageManagerName = workspace.getConfiguration('npm', folder.uri).get<string>('packageManager', 'npm');
+
+	if (packageManagerName === 'auto') {
+		const { name, multiplePMDetected } = await findPreferredPM(folder.uri.fsPath);
+		packageManagerName = name;
+
+		if (multiplePMDetected) {
+			const multiplePMWarning = localize('npm.multiplePMWarning', 'Found multiple lockfiles. Using {0} as the preferred package manager.', packageManagerName);
```
Comment (alexr00) [inline extensions/npm/src/tasks.ts:119]: Since each `folder` can have multiple PMs, we should include the folder's fspath in this message.

**Comment by alexr00** [comment]
@shivangg thanks for your patience! I made a some changes to be consistent with the rest of the codebase and for some cleanup. I also have two changes requested from you.

**Review by alexr00 — APPROVED** [review]
I addressed the last changes I wanted so that this can be merged.

**Comment by connor4312** [comment]
@alexr00 with this do you think it's reasonable to expose a command for extensions to get the package manager? Previously in js-debug I was using the [value of the setting](https://github.com/microsoft/vscode-js-debug/blob/f5336ee20077e2df077d1a42c401f445f68cabf0/src/ui/getRunScriptCommand.ts#L10-L11), but this no longer works and the 'auto' resolution is not trivial enough I would want to copy-paste it into js-debug.

image

**Comment by alexr00** [comment]
@connor4312 exposing a command to get the package manager seems perfectly reasonable. https://github.com/microsoft/vscode/issues/109071

**Comment by alexr00** [comment]
Reverted, we decided not to take on the additional dependencies.

**Comment by nickmccurdy** [comment]
Will this feature still be implemented or are you looking for another implementation with less dependencies?

**Comment by alexr00** [comment]
We have no plans to implement it in any specific time frame, but since it's on the backlog we still want to implement it eventually. You can see our [roadmap](https://github.com/microsoft/vscode/wiki/Roadmap) for the kinds of things we're planning on working on.

**Comment by nickmccurdy** [comment]
In that case, would it be helpful if I recreated this pull request with fewer/no dependencies?

**Comment by alexr00** [comment]
I have re-merged this PR with https://github.com/microsoft/vscode/commit/bea76730f10663abc91d840600c679f5b6fe60ac

**Pull request merged**